### DISCUSSION TITLE
Remove inherited_resources

### DISCUSSION THREAD

**Discussion by timoschilling** [activeadmin/activeadmin#7869] [Ideas]
`inherited_resources` will support Rails 4.2 for us, see https://github.com/activeadmin/activeadmin/pull/3193#issuecomment-62421649 for details.

To support Rails 5.0 in the future we need to remove `inherited_resources` and need to implement some features of `inherited_resources` in our base controller.

NOTE: This ticket is only for the replacement of `inherited_resources`, please don't post `inherited_resources` bugs and not no other off topics here!
And please no `+1` posts, you don't need to need to do that, it is a done thing, there is no way around on the long term.

**Comment by Fire-Dragon-DoL** [comment]
Are there any chance you guys can take over the development (and the upgrade) of inherited_resources (and keep it as a separate gem, which provides a double bonus). I don't use inherited_resources myself but I definitely agree that responders and respond_with is not a full replacement for what inherited_resources does, and considering how important is this for activeadmin, it can be option (also code will adapt to activeadmin requirements, which is quite good).

**Comment by jaredbeck** [comment]
Today, when asked to release a version of `inherited_resources` that supports rails 5, Rafael França said:

> .. for more than 4 years this gem is deprecated .. there is no motivation 
> by any of the maintainers to work on this gem.  .. I don't want to work on it.

Of course Timo is already aware of this, as he stated above, in 2014. I just thought the quote was interesting and shows the urgency of this issue.

**Comment by varyonic** [comment]
@Fire-Dragon-DoL Your [wish is granted](https://github.com/activeadmin/inherited_resources/pull/420/commits/849eecd13dd699c2528878165b87b992fc083b88)  For the curious, I believe inherited_resources was the inspiration and heart of ActiveAdmin, see Greg's [original repo](https://github.com/gregbell/inherited_views)

**Comment by Fire-Dragon-DoL** [comment]
!!!!
@varyonic thanks, that's a great surprise and comes with great surprise, making me extremely happy!

Good luck with the maintenance, and good job for all the work on ActiveAdmin!

**Comment by varyonic** [comment]
@timoschilling Can we close for now?  If someone has a PR to remove in future then great, but you already know how difficult that is.

**Comment by deivid-rodriguez** [comment]
I think it's good to keep this opened. It's something we want to do, so why not let contributors know about it with an opened issue? Maybe apply some enhacement / todo / help_wanted labels.

**Comment by timoschilling** [comment]
@varyonic I agree with @deivid-rodriguez

**Comment by kiddrew** [comment]
Is there any plan to move on from inherited_resources at this point? It's a relatively minor annoyance, but having BaseController inherit from ApplicationController through inherited_resources really loads up ApplicationController with a lot of logic that is specific to ActiveAdmin. Before hooks, etc. I know that before hooks can be done in the AA initializer, but that feels like a hack.

**Comment by Fivell** [comment]
@kiddrew  https://github.com/activeadmin/activeadmin/issues/3604#issuecomment-274981487 is actual (nothing changed since that time)

**Comment by javierjulio** [comment]
I figure removing this dependency would be too major as we likely use a lot from InheritedResources. If it's doable though I would be open to that and accept PRs. I'll move this to discussions for now instead of closing. Thank you.

**Reply by Fire-Dragon-DoL** [reply]
The core issue is that inherited resources effectively acts also as public API. ActiveAdmin doesn't have a formal public API beside what's  because available on activeadmin.info, it's made of many pieces and what's only on the website would be insufficient, so removing inherited resources would break part of that public API.

Thought from this perspective, there might be no reason to remove InheritedResources.

A lot of the value that activeadmin provides is in pre-built CRUD operations and UI, customizing this is obviously highly important.
In the end, InheritedResources is a bit dated, but it seems to align well with some of the values activeadmin provides.

I went through inherited_resources codebase and I was surprised it's not huge, it's possible to modernize it (and I'm not sure if it's even necessary).

I have a 10+ years old (I don't remember when it started) rails application where the core business value is provided by ActiveAdmin and even though the many issues over the years, the value that ActiveAdmin provides is in shortening the code I have to type for a set of standard CRUD operations across many tables. This obviously aligns with the famous saying "most applications are glorified CRUD forms", so there is a lot of value to provide on this side.

All of this to say, I changed my mind, I would actually recommend "dipping into" inherited resources, modernizing it a little bit if necessary, and expand the API. Even documenting it in the activeadmin website to formally provide a way to extend controllers (if you look at the websites, there is only the all-or-nothing `controller do` block documented).

Finally, all of  this is my opinion, I'm not a maintainer, so I'm just providing a perspective, it should be absolutely ignored if deemed irrelevant

**Reply by javierjulio** [reply]
Yes, that's right, it does act as a public API. Removing it would be simply too large of a task as it has its own set of features and tests. I'm happy to accept improvements there if anyone has anything. For now, I think long term, it's just about maintaining both projects as they are. For that I'll close this out as it's not realistic.

**Reply by Fire-Dragon-DoL** [reply]
I dipped into the code and its fairly simple. Simple code is helpful, very happy with the decision of keeping it

**Comment by javierjulio** [comment]
Removing this dependency would be simply too large of a task as it has its own set of features and tests. I'm happy to accept improvements there if anyone has anything. For now, I think long term, it's just about maintaining both projects as they are and to ensure continued compatibility (e.g. current Rails version).
